File: solution/2976. Minimum Cost to Convert String I/2976.cpp
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        vector<vector<long>> dist(26, vector<long>(26, INT_MAX));
        for (int i = 0; i < cost.size(); ++i) {
            int a = original[i] - 'a', b = changed[i] - 'a';
            dist[a][b] = min(dist[a][b], 1L * cost[i]);
        }
        
        for (int i = 0; i < 26; ++i)
            dist[i][i] = 0;
        
        for (int k = 0; k < 26; ++k) {
            for (int i = 0; i < 26; ++i) {
                for (int j = 0; j < 26; ++j) {
                    if (dist[i][k] == INT_MAX || dist[k][j] == INT_MAX) continue;
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
        
        long res = 0;
        for (int i = 0; i < target.size(); ++i) {
            int a = source[i] - 'a', b = target[i] - 'a';
            if (dist[a][b] == INT_MAX) return -1;
            res += dist[a][b];
        }
        return res;
    }
};

File: solution/2976. Minimum Cost to Convert String I/2976.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        dist = [[float('inf')] * 26 for _ in range(26)]
        
        for i in range(len(cost)):
            a, b = ord(original[i]) - ord('a'), ord(changed[i]) - ord('a')
            dist[a][b] = min(dist[a][b], cost[i])
        
        for i in range(26):
            dist[i][i] = 0
        
        for k in range(26):
            for i in range(26):
                for j in range(26):
                    if dist[i][k] == float('inf') or dist[k][j] == float('inf'):
                        continue
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
        
        res = 0
        for i in range(len(target)):
            a, b = ord(source[i]) - ord('a'), ord(target[i]) - ord('a')
            if dist[a][b] == float('inf'):
                return -1
            res += dist[a][b]
        return res
    
